{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a252d23c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6abdaafe",
   "metadata": {},
   "source": [
    "## Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd49823e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input mutation files\n",
    "\n",
    "# annotated mutation call positions wrt to the EB reference \n",
    "fname_mutations_wild_type = '../../resources/run_workflow/results_cluster/wild_type/all_mutations.annotated.corrected.csv'\n",
    "fname_mutations_ko = '../../resources/run_workflow/results_cluster/dicer_KO/all_mutations.annotated.corrected.csv'\n",
    "fname_mutations_overexpressed = '../../resources/run_workflow/results_cluster/dicer_overexpression/all_mutations.annotated.corrected.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "47d19df1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wild_type = pd.read_csv(fname_mutations_wild_type)\n",
    "df_wild_type = df_wild_type.drop(['Unnamed: 0'], axis =1)\n",
    "df_wild_type['genotype'] = 'wild_type'\n",
    "\n",
    "df_ko = pd.read_csv(fname_mutations_ko)\n",
    "df_ko = df_ko.drop(['Unnamed: 0'], axis =1)\n",
    "df_ko['genotype'] = 'dicer_ko'\n",
    "\n",
    "df_overexpressed = pd.read_csv(fname_mutations_overexpressed)\n",
    "df_overexpressed = df_overexpressed.drop(['Unnamed: 0'], axis =1)\n",
    "df_overexpressed['genotype'] = 'dicer_overexpressed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "210791a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataframe with all mutations from all samples\n",
    "df = pd.concat([df_ko, df_overexpressed, df_wild_type])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "528ebc06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO',\n",
       "       'genotype', 'replicate', 'passage', 'Fvar', 'Rvar', 'Ftot', 'Rtot',\n",
       "       'Pval', 'Qval', 'Freq1', 'Freq2', 'Freq3', 'Post1', 'Post2', 'Post3',\n",
       "       'RefCodon', 'AltCodon', 'RefAminoAcid', 'AltAminoAcid', 'CodonPosition',\n",
       "       'SNPCodonPosition', 'AminoAcidChange', 'IsSynonymous', 'IsTransition',\n",
       "       'IsGenic', 'IsPseudo', 'LocusTag', 'Gene', 'Note', 'Inference',\n",
       "       'Product', 'ProteinID', 'Comments', 'VariantType', 'FeatureType'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b91492ba",
   "metadata": {},
   "source": [
    "## Add information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0342c196",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add information\n",
    "df['Freq1'] = pd.to_numeric(df['Freq1'], errors='coerce')\n",
    "df['Freq2'] = pd.to_numeric(df['Freq2'], errors='coerce')\n",
    "df['Freq3'] = pd.to_numeric(df['Freq3'], errors='coerce')\n",
    "\n",
    "df['n_reads_var'] = df['Rvar'] + df['Fvar']\n",
    "df['coverage'] = df['Rtot'] + df['Ftot']\n",
    "df['frequency'] = df['n_reads_var'] / df['coverage']\n",
    "df['Frq_ave'] = df[['Freq1','Freq2','Freq3']].mean(axis=1)\n",
    "\n",
    "df['passage'] = df['passage'].str.split('_').str[-1].astype('float')\n",
    "\n",
    "df['position'] = df['POS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b3bbcc",
   "metadata": {},
   "source": [
    "## Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "dfbec31a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Post-processing filtering of mutation calling\n",
    "\n",
    "# filter out where Frq_ave == 0.0 \n",
    "# that is something unexpected happening in ShoRAH which is due to the super high coverage\n",
    "df = df[df['Frq_ave']!=0]\n",
    "\n",
    "# strand bias test \n",
    "df = df[df['Pval']>=0.05]\n",
    "\n",
    "# minimum read support\n",
    "minimum_read_support = 10 \n",
    "df = df[df['n_reads_var']>=minimum_read_support]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0570fa56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>genotype</th>\n",
       "      <th>replicate</th>\n",
       "      <th>...</th>\n",
       "      <th>Product</th>\n",
       "      <th>ProteinID</th>\n",
       "      <th>Comments</th>\n",
       "      <th>VariantType</th>\n",
       "      <th>FeatureType</th>\n",
       "      <th>n_reads_var</th>\n",
       "      <th>coverage</th>\n",
       "      <th>frequency</th>\n",
       "      <th>Frq_ave</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>190</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>15.24810</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=28;Rvar=89;Ftot=15146;Rtot=64088;Pval=0.4...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>117</td>\n",
       "      <td>79234</td>\n",
       "      <td>0.001477</td>\n",
       "      <td>0.001200</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>219</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.40133</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=15;Rvar=28;Ftot=21761;Rtot=38660;Pval=1.0...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>43</td>\n",
       "      <td>60421</td>\n",
       "      <td>0.000712</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>222</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4.53375</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=21;Rvar=53;Ftot=22546;Rtot=35907;Pval=0.1...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>74</td>\n",
       "      <td>58453</td>\n",
       "      <td>0.001266</td>\n",
       "      <td>0.000367</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>246</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4.55477</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=13;Rvar=8;Ftot=27030;Rtot=22623;Pval=0.67...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>21</td>\n",
       "      <td>49653</td>\n",
       "      <td>0.000423</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>292</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>20.86190</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=82;Rvar=33;Ftot=35315;Rtot=17916;Pval=0.4...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>115</td>\n",
       "      <td>53231</td>\n",
       "      <td>0.002160</td>\n",
       "      <td>0.000900</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>293</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4.71426</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=46;Rvar=12;Ftot=35629;Rtot=18124;Pval=0.1...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>58</td>\n",
       "      <td>53753</td>\n",
       "      <td>0.001079</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>331</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>24.31800</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=15;Rvar=18;Ftot=40664;Rtot=29252;Pval=0.2...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>33</td>\n",
       "      <td>69916</td>\n",
       "      <td>0.000472</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>331</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>25.74030</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=1635;Rvar=795;Ftot=40664;Rtot=29252;Pval=...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>2430</td>\n",
       "      <td>69916</td>\n",
       "      <td>0.034756</td>\n",
       "      <td>0.009100</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>354</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4.48062</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=63;Rvar=47;Ftot=45154;Rtot=35426;Pval=0.9...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>110</td>\n",
       "      <td>80580</td>\n",
       "      <td>0.001365</td>\n",
       "      <td>0.000700</td>\n",
       "      <td>354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AF014388</td>\n",
       "      <td>412</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>17.25070</td>\n",
       "      <td>PASS</td>\n",
       "      <td>Fvar=148;Rvar=134;Ftot=46470;Rtot=40694;Pval=0...</td>\n",
       "      <td>dicer_ko</td>\n",
       "      <td>replicate_d</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>SNP</td>\n",
       "      <td>inter_genic</td>\n",
       "      <td>282</td>\n",
       "      <td>87164</td>\n",
       "      <td>0.003235</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM  POS ID REF ALT      QUAL FILTER  \\\n",
       "0   AF014388  190  .   A   G  15.24810   PASS   \n",
       "2   AF014388  219  .   A   T   4.40133   PASS   \n",
       "3   AF014388  222  .   A   G   4.53375   PASS   \n",
       "4   AF014388  246  .   T   C   4.55477   PASS   \n",
       "8   AF014388  292  .   A   T  20.86190   PASS   \n",
       "9   AF014388  293  .   C   T   4.71426   PASS   \n",
       "12  AF014388  331  .   -   C  24.31800   PASS   \n",
       "13  AF014388  331  .   -   T  25.74030   PASS   \n",
       "14  AF014388  354  .   A   G   4.48062   PASS   \n",
       "15  AF014388  412  .   T   C  17.25070   PASS   \n",
       "\n",
       "                                                 INFO  genotype    replicate  \\\n",
       "0   Fvar=28;Rvar=89;Ftot=15146;Rtot=64088;Pval=0.4...  dicer_ko  replicate_d   \n",
       "2   Fvar=15;Rvar=28;Ftot=21761;Rtot=38660;Pval=1.0...  dicer_ko  replicate_d   \n",
       "3   Fvar=21;Rvar=53;Ftot=22546;Rtot=35907;Pval=0.1...  dicer_ko  replicate_d   \n",
       "4   Fvar=13;Rvar=8;Ftot=27030;Rtot=22623;Pval=0.67...  dicer_ko  replicate_d   \n",
       "8   Fvar=82;Rvar=33;Ftot=35315;Rtot=17916;Pval=0.4...  dicer_ko  replicate_d   \n",
       "9   Fvar=46;Rvar=12;Ftot=35629;Rtot=18124;Pval=0.1...  dicer_ko  replicate_d   \n",
       "12  Fvar=15;Rvar=18;Ftot=40664;Rtot=29252;Pval=0.2...  dicer_ko  replicate_d   \n",
       "13  Fvar=1635;Rvar=795;Ftot=40664;Rtot=29252;Pval=...  dicer_ko  replicate_d   \n",
       "14  Fvar=63;Rvar=47;Ftot=45154;Rtot=35426;Pval=0.9...  dicer_ko  replicate_d   \n",
       "15  Fvar=148;Rvar=134;Ftot=46470;Rtot=40694;Pval=0...  dicer_ko  replicate_d   \n",
       "\n",
       "    ...  Product  ProteinID  Comments  VariantType  FeatureType  n_reads_var  \\\n",
       "0   ...        .          .         .          SNP  inter_genic          117   \n",
       "2   ...        .          .         .          SNP  inter_genic           43   \n",
       "3   ...        .          .         .          SNP  inter_genic           74   \n",
       "4   ...        .          .         .          SNP  inter_genic           21   \n",
       "8   ...        .          .         .          SNP  inter_genic          115   \n",
       "9   ...        .          .         .          SNP  inter_genic           58   \n",
       "12  ...        .          .         .          SNP  inter_genic           33   \n",
       "13  ...        .          .         .          SNP  inter_genic         2430   \n",
       "14  ...        .          .         .          SNP  inter_genic          110   \n",
       "15  ...        .          .         .          SNP  inter_genic          282   \n",
       "\n",
       "    coverage  frequency   Frq_ave  position  \n",
       "0      79234   0.001477  0.001200       190  \n",
       "2      60421   0.000712  0.000067       219  \n",
       "3      58453   0.001266  0.000367       222  \n",
       "4      49653   0.000423  0.000033       246  \n",
       "8      53231   0.002160  0.000900       292  \n",
       "9      53753   0.001079  0.000133       293  \n",
       "12     69916   0.000472  0.000167       331  \n",
       "13     69916   0.034756  0.009100       331  \n",
       "14     80580   0.001365  0.000700       354  \n",
       "15     87164   0.003235  0.003000       412  \n",
       "\n",
       "[10 rows x 48 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "19622f4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('annotated_mutations_of_all_samples_after_filtering.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4abea852",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
